### PR TITLE
feat(config): Adds enable_default_collections flag to do not initialize collections by default

### PR DESCRIPTION
**Pull request by M4tteoP**
Hello,
I tried to address https://github.com/coreruleset/coreruleset/issues/3068, following the decision taken during [Jan 16 (issue chat meeting)](https://github.com/coreruleset/coreruleset/issues/3037#issuecomment-1384488207) about relying on a flag to initialize collections for the plugins, keeping it off by default.
Not sure the implementation is as you would expect, happy to have any guidance about it.
Thanks!

Tentatively closes https://github.com/coreruleset/coreruleset/issues/3068

### CODE REVIEW

File: rules/REQUEST-901-INITIALIZATION.conf [hunk 1]
```diff
@@ -281,20 +281,33 @@ SecAction \
 # -=[ Initialize collections ]=-
 #
 # Create both Global and IP collections for rules to use.
-# There are some CRS rules that assume that these two collections
-# have already been initiated.
+# Some plugins assume that these two collections have already
+# been initiated.
```
Comment (theseion) [inline rules/REQUEST-901-INITIALIZATION.conf:285]:
```suggestion
# been initialized.
```

File: rules/REQUEST-901-INITIALIZATION.conf [hunk 2]
```diff
@@ -281,20 +281,33 @@ SecAction \
...
 #
 
-SecRule REQUEST_HEADERS:User-Agent "@rx ^.*$" \
+# Default Enable Collections (rule 900130 in crs-setup.conf)
```
Comment (theseion) [inline rules/REQUEST-901-INITIALIZATION.conf:288]:
```suggestion
# Disable collection initialisation by default (see rule 900130 in crs-setup.conf)
```

File: rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf [hunk 3]
```diff
@@ -544,10 +544,10 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
 # This enables an attacker to craft a string that will be delivered in a form that a browser will execute as script
 # while being ignored by input filters.
 #
-# This rule looks for start tag sequene that looks like "<...>" (checks fo hex and plain to be sure).
+# This rule looks for start tag sequence that looks like "<...>" (checks for hex and plain to be sure).
```
Comment (theseion) [inline rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf:547]:
```suggestion
# This rule looks for a start tag sequence that looks like "<...>" (checks for hex and plain to be sure).
```

**Review by theseion — CHANGES_REQUESTED** [review]
Looks solid to me, but I'm a bit fuzzy on the details of setup.

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
I've asked @dune73 to review as well.

**Comment by dune73** [comment]
Thank you for picking this up, @M4tteoP. It's good somebody is taking care of it.

The way you do this, probably works, but I would like to see this optimized.

* We do not need rule 901315. Honestly, if the count of this variable is 0, then it's not set. So we can use the var counter as the first rule in a chain.
* 901318 is nice, but no need to make this into a separate rule. Integrate into the single rule that sets everything.
* 901321, here you check for the existence of the variable, then you chain a rule that checks for the value of the variable, then you chain a rule that creates the UA_Hash and initiates the collections without creating a ua-hash variable that we do not need anyways.

Well, maybe we need to think about the use case and keeping the ua-hash variable around, but I am confident the way ModSec works with variables, you do not need the key anymore once it's initialized.

Sorry this makes your PR much more complicated, but now that we touch it, we should also make it simpler and more performant.

**Comment by M4tteoP** [comment]
Thanks for the guidance! PTAL 🙏

**Review by dune73 — APPROVED** [review]

**Comment by dune73** [comment]
Thank you. Ready to fly from my side.

**Comment by theseion** [comment]
Looks good to me too. One thing I noticed: why do we need two chained rules? We're checking `... @eq 1` anyway, so do we really need to check that the variable is set first?

**Comment by M4tteoP** [comment]
Done, I did some little tests, and it seems that it is working as expected, but the lint is complaining about relying on a unset variable 🤔:
```
::error  There are one or more unset TX variables.
::error  file=rules/REQUEST-901-INITIALIZATION.conf,line=289,endLine=289,title=unset TX variable: TX variable 'ENABLE_DEFAULT_COLLECTIONS' not set / later set (VAR)
```

**Comment by theseion** [comment]
And it didn't complain before? It should have complained before, I think. Well, you just have to set the variable explicitly. We introduced this linter step for a reason :)

**Comment by dune73** [comment]
How do you mean before? It's a new variable and the initial PR did set the variable explicitly.

I think we've been in this discussion before. Adding a SecRule just to please the linter seems wrong to me. This is a case where the variable is optional and the rule treats that just fine.

@airween: What do you think?

**Comment by dune73** [comment]
Thanks for chiming in @airween. I remember the discussion lively and also the decision to enforce variable initialization always. However, I am no longer familiar with the use case and my gut feeling tells me the use case was less clear than the one here, so maybe we can return on the question.

Here, we have a situation where initialization adds an overhead of an additional SecRule that is not needed otherwise. Given most people are likely to run without collections, the initialization of this optional variable seems entirely futile to me.

But this is a special use case. Generally, we should enforce variable initialization and we have discovered several bugs where the vars have not been initialized. So the check makes a lot of sense in 90% of the cases.

For me the question is what to do with the remaining cases. And the one here falls into this category:
* The variable is optional
* Most people will leave it off (or unitialized if they can)
* The rules using the variables are diligently created and reviewed. It works just fine when the variable is not initialized.

I think in such a case, we should be able to make an exception to the general rule.

I see two ways:
* Hard-Code the optional variable names into the linter
* Add a special format comment read by the linter that allows to pass it the name of a variable that does not have to be initialized.

If I can convince you that the overhead of the unnecessary initialization should be avoided.

**Comment by airween** [comment]
> Here, we have a situation where initialization adds an overhead of an additional SecRule that is not needed otherwise.

May be I misunderstand something, but I think the [enforce_bodyproc_urlencoded](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/rules/REQUEST-901-INITIALIZATION.conf#L225) and [crs_validate_utf8_encoding](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/rules/REQUEST-901-INITIALIZATION.conf#L234) are the same as this new one. We can dispense with the initialization of variables (even just we use it only one place), but then the rule set won't consistent.

> So the check makes a lot of sense in 90% of the cases.

How can we decide which cases makes sense and which don't? (My algorithm asks :))

> * The variable is optional

Same question as above - what makes a variable optional?

> * Most people will leave it off (or unitialized if they can)

The `rules-check` script prevents it :)

> * The rules using the variables are diligently created and reviewed. It works just fine when the variable is not initialized.

Is there any way to check that the written rule works as we expected in **each** cases?


Anyway,
> I see two ways:
>
> * Hard-Code the optional variable names into the linter

I should avoid that.

> * Add a special format comment read by the linter that allows to pass it the name of a variable that does not have to be initialized.

I thought the special comment too, or we can use some name convention for these optional variables. Eg. the `OPT_` prefix, like `OPT_ENABLE_DEFAULT_COLLECTIONS`. Linter can recognize this prefix, and can it can turn a blind eye.

If I modify the linter, should we remove the mentioned two rules? I would insist the consistent usage of variables.

**Comment by theseion** [comment]
The discussion happened on Slack: https://github.com/coreruleset/project-chat-archive/blob/master/chat-archive-2022-11-21.md.
There's a related GH issue too: https://github.com/coreruleset/coreruleset/issues/2990.

The discussion then was pretty much identical :). And I think we've discussed this before too: why don't we just initialize the variable to 0? We have the rule in the setup file, it's just commented out. Or we do what @airween ded for `crs_validate_utf8_encoding` and set the variable if it is unset (https://github.com/coreruleset/coreruleset/pull/2802/files), which is, I think, what @M4tteoP had written in the first place :).

I'm with @airween on this one, if we allow variables to be uninitialized, however we do it, at some point we will be in a similar situation, where we have a bunch of `OPT_*` variables that are no longer in use and we'll spend unnecessary time figuring that out.

**Comment by dune73** [comment]
Thanks for the link.

My reasoning is purely around performance: Why execute an unnecessary SecRule to initialize something that 90% of users do not need.

A SecRule directive has an overhead of 0.5 microseconds. All CRS systems combined serve many, many billions of pages a day. This adds up. And what we get in return is convenience and a cleaner code base. I think this is not worth it. 

I know I am probably in a minority position with this discussion again. I just do not understand why my performance argument is so irrelevant to you.

**Comment by theseion** [comment]
I do get the argument for performance. Do the 0.5 microseconds also apply to chained rules? Because we do _check_ the variable either way.

Another idea: @airween, would it be possible to read `crs-setup.conf`, uncomment the commented rules and run the check then? Because then the optional variables _would_ have been set.

**Comment by airween** [comment]
> Another idea: @airween, would it be possible to read `crs-setup.conf.example`, uncomment the commented rules and run the check then? Because then the optional variables _would_ have been set.

When I started to work on the `rules-check`, I had a similar idea - but I disagreed it, because I thought that leads to misunderstandings.

I you guys agree that this would be a useful way, I can take a look it how could I do that.

**Comment by dune73** [comment]
I think @theseion's proposal could be a useful way of dealing with the problem.

Defining where such a commented-out initalization is permitted and where it's not would be part of the review process I guess. Also possible to add guidance to the contributing.md doc.

And I agree that an individual rule adds but little. But this is how cruft piles up.

**Comment by M4tteoP** [comment]
Small indentations and actions ordering fixes, but after https://github.com/coreruleset/coreruleset/pull/3161, the linter is happy!

**Comment by theseion** [comment]
Why was this PR reopened @airween?

**Comment by airween** [comment]
> Why was this PR reopened @airween?

because we haven't finished yet. I'm afraid I accidentally closed it.

**Comment by theseion** [comment]
Let me know when you want a review.

**Comment by M4tteoP** [comment]
This PR was on hold because it is requiring the commented-out initialization, I think now it is ready to have a second check. Thanks!

**Review by dune73 — APPROVED** [review]

**Comment by dune73** [comment]
LGTM. Thanks.

**Comment by dune73** [comment]
Let's merge this. Thank you very much for the contribution @M4tteoP.

**Pull request merged**